### PR TITLE
⚠️ Bump controller-runtime 0.13.1=>0.14.1

### PR DESCRIPTION
**Pull request by aniruddha2000**
Signed-off-by: Aniruddha Basak <codewithaniruddha@gmail.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #7671 

cc @kubernetes-sigs/cluster-api-release-team

### CODE REVIEW

**Review by ykakarap — COMMENTED** [review]
Looks like this version of controller-runtime introduced some breaking changes.
There are a few lint issue that need to be addressed. Running the linter locally I see the following issues:
- The `ValidateLabelSelector` function signature changed. Passing an empty options object (`metav1validation.LabelSelectorValidationOptions{}`) as the second argument, at every use, should be enough to fix that.
- sets.String seems to be deprecated and now the recommendation seems to be to use generics to create a set. Doing that switch is out of scope for this PR. Let's do it in a separate PR. So for now, I would suggest adding an exception in the linter for the scope of this PR and address it in a follow up PR.
- controller-runtime's `client.Client` interface changes and the `dryrun.Client` struct no longer implements the interface correctly.  I would suggest adding the required functions to the dryrun.Client struct should be enough for this PR. I will need more time to see if we can get away with an "empty" implmentation as the dryrun client does not operate on subresources.

File: .golangci.yml [hunk 1]
```diff
@@ -268,6 +268,10 @@ issues:
     - staticcheck
     text: "SA1019: in.(.+) is deprecated"
     path: .*(api|types)\/.*\/conversion.*\.go$
+  - linters:
+    - staticcheck
+    text: "SA1019: is deprecated in sets.String"
+    path: .*(cmd|controllers|controlplane|internal)\/.*\/*\.go$
```
Comment (aniruddha2000) [inline .golangci.yml:274]: @ykakarap I am having issue finding why this regex is not catching the path

Comment (sbueringer) [reply .golangci.yml:274]: path is okay (although I would drop the `\/*` in the middle. The problem is the text. The following should work:
```
  - linters:
    - staticcheck
    text: "SA1019: sets.String is deprecated"
    path: .*(cmd|controllers|controlplane|internal)\/.*\.go$
```

Comment (sbueringer) [reply .golangci.yml:274]: Looks like an additional `make generate-modules` is required as well

Comment (sbueringer) [reply .golangci.yml:274]: Then the final finding should be 
```
test/framework/namespace_helpers.go:147:31: Error return value of `eventInformer.AddEventHandler` is not checked (errcheck)
        eventInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
```

Let's write the returned error into `err` + verify it's nil via: 
```go
Expect(err).NotTo(HaveOccurred())
```

**Comment by sbueringer** [comment]
Changes lgtm pending what is mentioned in this thread: https://github.com/kubernetes-sigs/cluster-api/pull/7906#discussion_r1069002758

**Comment by aniruddha2000** [comment]
Lint passed but seems there are a lot of complicated breaking changes now 😵‍💫

**Comment by sbueringer** [comment]
@aniruddha2000 I can take a look at it and then open a PR against your branch if you want to

**Comment by aniruddha2000** [comment]
@sbueringer sure!

**Comment by sbueringer** [comment]
> @sbueringer sure!

Will do. Looks like some of the failures are related to some things I identified once I got a chance to catch up with contorller-runtime yesterday (https://github.com/kubernetes-sigs/cluster-api/issues/7925)

**Comment by sbueringer** [comment]
@aniruddha2000 PR is open here: https://github.com/aniruddha2000/cluster-api/pull/10
Maybe just merge it into your PR, so my change will end up on this PR

Then we can take it from there.

The problem was essentially that fake client now supports indexes and we have to set them on the fake client explicitly

File: .golangci.yml (hunk 1 above)
Comment (sbueringer) [reply .golangci.yml]: I'll take care of the follow-up.

**Comment by sbueringer** [comment]
Thx!
Can you please squash the commits?

/lgtm

/assign @fabriziopandini @ykakarap

File: internal/controllers/machine/machine_controller_test.go [hunk 2]
```diff
@@ -723,7 +724,9 @@ func TestReconcileRequest(t *testing.T) {
 				&tc.machine,
 				builder.GenericInfrastructureMachineCRD.DeepCopy(),
 				&infraConfig,
-			).Build()
+			).
+				WithIndex(&corev1.Node{}, index.NodeProviderIDField, index.NodeByProviderID).
+				Build()
```
Comment (ykakarap) [inline internal/controllers/machine/machine_controller_test.go:729]: nit: can you please fix the indentation.
(same for the below in the file)

**Review by ykakarap — COMMENTED** [review]
LGTM pending squash.

Just minor nits.

Thank you @aniruddha2000 for all the work! :rocket:

File: .golangci.yml [hunk 3]
```diff
@@ -268,6 +268,10 @@ issues:
     - staticcheck
     text: "SA1019: in.(.+) is deprecated"
     path: .*(api|types)\/.*\/conversion.*\.go$
+  - linters:
+    - staticcheck
+    text: "SA1019: sets.String is deprecated"
```
Comment (fabriziopandini) [inline .golangci.yml:273]: @sbueringer should we open an issue to fix the usage of this deprecated type

**Review by fabriziopandini — COMMENTED** [review]
/lgtm

**Comment by aniruddha2000** [comment]
Squash done

File: .golangci.yml (hunk 3 above)
Comment (sbueringer) [reply .golangci.yml:273]: Yup it's a part of https://github.com/kubernetes-sigs/cluster-api/issues/7925

File: docs/book/src/tasks/experimental-features/runtime-sdk/implement-extensions.md [hunk 4]
```diff
@@ -46,8 +46,8 @@ import (
 	"github.com/spf13/pflag"
 	cliflag "k8s.io/component-base/cli/flag"
```
Comment (sbueringer) [inline docs/book/src/tasks/experimental-features/runtime-sdk/implement-extensions.md:47]: @aniruddha2000 can you please drop this diff from the PR?

Comment (sbueringer) [reply docs/book/src/tasks/experimental-features/runtime-sdk/implement-extensions.md:47]: It seems to be the same as was already merged in https://github.com/kubernetes-sigs/cluster-api/pull/7912/files

Comment (sbueringer) [reply docs/book/src/tasks/experimental-features/runtime-sdk/implement-extensions.md:47]: Hm. Took a closer look. Not sure why the GitHub UI is showing the diff. I'll merge it should be fine.

**Comment by RainbowMango** [comment]
Hi @kubernetes-sigs/cluster-api-release-team,

Could you please tag a new patch release to include this?

We([Karmada project](https://github.com/karmada-io/karmada)) are going to bump Kubernetes(v1.26) and controller-runtime(v0.14.1), but we are blocked by cluster-api because we are using cluster-api@v1.0.1.

I checked the latest release v1.3.2 which does not include this patch.
I'd like a new tagged release from cluster-api, like v1.3.3? 

```
# sigs.k8s.io/cluster-api/api/v1beta1
vendor/sigs.k8s.io/cluster-api/api/v1beta1/machineset_types.go:180:83: not enough arguments in call to metav1validation.ValidateLabelSelector
	have (*"k8s.io/apimachinery/pkg/apis/meta/v1".LabelSelector, *field.Path)
	want (*"k8s.io/apimachinery/pkg/apis/meta/v1".LabelSelector, validation.LabelSelectorValidationOptions, *field.Path)
```

**Comment by killianmuldoon** [comment]
@RainbowMango according to the release timeline - https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/release/release-1.4.md#timeline - there should be release of v1.3.3 this week.

**Comment by furkatgofurov7** [comment]
> @RainbowMango according to the release timeline - https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/release/release-1.4.md#timeline - there should be release of v1.3.3 this week.

@killianmuldoon this was merged to main and not back-ported to release-1.3 branch AFAIU, so I would not expect to have these changes in the new patch release (v1.3.3).

**Comment by RainbowMango** [comment]
OK, that makes sense. For now, I can use a pseudo version from the main branch(like v1.3.0-rc.0.0.20230127161026-14ffcb25bbf2), and change it to v1.4.0 after v1.4.0 is released.

Thanks @killianmuldoon @furkatgofurov7. That really helps.

**Comment by sbueringer** [comment]
For clarification. We consider bumping controller runtime and k8s.io/* dependencies breaking changes. Thus this won't be backported.

**Pull request merged**